File: qqbot-client/src/main/java/com/mieai/qqbot/client/QqRichMessageRequest.java
package com.mieai.qqbot.client;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/** Extensible rich-message envelope shared by all supported QQ message targets. */
public record QqRichMessageRequest(
        QqMessageTargetType targetType,
        String targetId,
        QqRichMessageKind kind,
        Map<String, Object> payload,
        Optional<String> replyMessageId,
        Optional<String> replyEventId,
        int messageSequence) {
    public QqRichMessageRequest {
        Objects.requireNonNull(targetType, "targetType must not be null");
        if (targetId == null || targetId.isBlank() || !targetId.equals(targetId.strip())
                || targetId.codePoints().anyMatch(Character::isWhitespace)) {
            throw new IllegalArgumentException("targetId is invalid");
        }
        Objects.requireNonNull(kind, "kind must not be null");
        payload = Map.copyOf(Objects.requireNonNull(payload, "payload must not be null"));
        if (payload.isEmpty()) throw new IllegalArgumentException("payload must not be empty");
        if (kind == QqRichMessageKind.KEYBOARD) validateKeyboardPayload(payload);
        Objects.requireNonNull(replyMessageId, "replyMessageId must not be null");
        Objects.requireNonNull(replyEventId, "replyEventId must not be null");
        if (messageSequence < 1) throw new IllegalArgumentException("messageSequence must be positive");
    }

    private static void validateKeyboardPayload(Map<String, Object> value) {
        if (!value.keySet().equals(java.util.Set.of("markdown", "keyboard"))
                || !(value.get("markdown") instanceof Map<?, ?> markdown)
                || !(value.get("keyboard") instanceof Map<?, ?> keyboard)
                || keyboard.isEmpty()) {
            throw new IllegalArgumentException("keyboard payload must contain markdown and keyboard objects");
        }
        Object content = markdown.get("content");
        Object template = markdown.get("custom_template_id");
        boolean hasContent = content instanceof String text && !text.isBlank();
        boolean hasTemplate = template instanceof String text && !text.isBlank();
        if (!hasContent && !hasTemplate) {
            throw new IllegalArgumentException("keyboard markdown must contain content or custom_template_id");
        }
    }
}

File: qqbot-client/src/main/kotlin/com/mieai/qqbot/client/QqRichMessageRequest.kt
package com.mieai.qqbot.client

/** Extensible rich-message envelope shared by all supported QQ message targets. */
class QqRichMessageRequest(
    val targetType: QqMessageTargetType,
    val targetId: String,
    val kind: QqRichMessageKind,
    payload: Map<String, Any>,
    val replyMessageId: String?,
    val replyEventId: String?,
    val messageSequence: Int,
) {
    val payload: Map<String, Any> = payload.toMap()

    init {
        require(targetId.isNotBlank() && targetId == targetId.trim() &&
            targetId.codePoints().noneMatch(Character::isWhitespace)) {
            "targetId is invalid"
        }
        require(payload.isNotEmpty()) { "payload must not be empty" }
        require(messageSequence >= 1) { "messageSequence must be positive" }
        if (kind == QqRichMessageKind.KEYBOARD) validateKeyboardPayload(payload)
    }

    private companion object {
        fun validateKeyboardPayload(value: Map<String, Any>) {
            val markdown = value["markdown"]
            val keyboard = value["keyboard"]
            require(value.keys == setOf("markdown", "keyboard") && markdown is Map<*, *> &&
                keyboard is Map<*, *> && keyboard.isNotEmpty()) {
                "keyboard payload must contain markdown and keyboard objects"
            }
            val content = markdown["content"]
            val template = markdown["custom_template_id"]
            val hasContent = content is String && content.isNotBlank()
            val hasTemplate = template is String && template.isNotBlank()
            require(hasContent || hasTemplate) {
                "keyboard markdown must contain content or custom_template_id"
            }
        }
    }
}
